README.md

```diff
@@ -1,91 +1,188 @@
-3. Update [context.ts](./src/types/context.ts) with the events that your plugin will fire on.
-4. Update [manifest.json](./manifest.json) with a proper description of your plugin.
-5. Update [plugin-inputs.ts](./src/types/plugin-inputs.ts) to match the `with:` settings in your org or repo level configuration.
+#### Environment Variables
 
-- Your plugin config should look similar to this:
+The `TELEGRAM_BOT_ENV` is a single JSON object that encapsulates all necessary environment variables for the bot's operation. It consists of four key sections: `botSettings`, `mtProtoSettings`, `storageSettings`, and `ubiquityOsSettings`.
 
-```yml
-plugins:
-  - name: hello-world
-    id: hello-world
-    uses:
-      - plugin: http://localhost:4000
-        with:
-          # Define configurable items here and the kernel will pass these to the plugin.
-          configurableResponse: "Hello, is it me you are looking for?"
-```
+You can set up your environment variables by using the provided utility script:
+
+- Run `yarn setup-env`, which prompts you to enter each value via the CLI. The values will be serialized and stored both locally and in your repository secrets.
 
-###### At this stage, your plugin will fire on your defined events with the required settings passed in from the kernel. You can now start writing your plugin's logic.
+- **GITHUB_PAT_TOKEN**: Create a classic personal access token (PAT) with the `repo` scope. Set the expiry to 24 hours. This token will be used to generate repository secrets for the environment variables and will be removed from `.env` after the secrets are saved.
+- **Account Permission**: The account in which the PAT is associated with _must_ be an `admin` of the repository to be able to save secrets this way. Visit your repository settings `telegram-bot` > `Collaborators & teams` to add the account as an admin first if needed.
 
-6. Start building your plugin by adding your logic to the [plugin.ts](./src/plugin.ts) file.
+The environment variables are stored in the following locations:
 
-## Testing a plugin
+- `.env` file: Required to run the `yarn sms-auth` command.
+- `.dev.vars` file: For the Cloudflare Worker instance.
+- **GitHub Secrets**: Used by the GitHub Actions workflow.
 
-### Worker Plugins
+##### Environment Variable Sections:
 
-- `yarn/bun worker` - to run the worker locally.
-- To trigger the worker, `POST` requests to http://localhost:4000/ with an event payload similar to:
+1. **botSettings**: Contains bot-specific settings like `TELEGRAM_BOT_TOKEN`, `TELEGRAM_BOT_WEBHOOK_SECRET`, etc.
+2. **mtProtoSettings**: Contains settings for the MTProto API like `TELEGRAM_APP_ID`, `TELEGRAM_API_HASH`, etc.
+3. **storageSettings**: Contains settings for the Supabase database like `SUPABASE_URL`, `SUPABASE_SERVICE_KEY`, etc.
+4. **ubiquityOsSettings**: Contains settings for the UbiquityOS Kernel like `APP_PRIVATE_KEY`, `APP_ID`, etc.
```

Comment:
for `rest.actions.createWorkflowDispatch` so this can dispatch it's own workflows. I believe you must use an app authenticated octokit instance and I don't believe the `authToken` being passed down is sufficient is it?

Reply:
Does it make sense to make the "telegram kernel" a separate github app? Or would that require more installations for partners? Or can we set it up once and let anyone use it?

Reply:
when the plugin receives `issues.opened` event, is it just forwarding the event to the workflow or is it doing any other work? 

If so, it doesn't make sense to just forward events when you can use workflows directly, for example you can just define both workflow and worker endpoints in the config, so there is not need for forwarding.

Reply:
As I understand, the initial webhook receiver for this system is a cloudflare worker. This is in case we need a fast response (like a telegram chat bot message) 

Since it is on cloudflare workers to start, I don't think it's possible to directly execute GitHub actions

Reply:
One worker, one entry and that entry is used by both payloads, we typecheck the payload and route it to it's destination.

Workflows are forwarded payloads from the worker.

All `proxyCallbacks` assigned to `issues.labeled` will run, e.g `workrooms` will forward the payload to the workflow entry which will run it's own `workflowProxyCallbacks` opening the room.

This feature does not require the worker to do anything except forward the payload. But other features may execute within the worker under the same event `issues.labeled` for example let's pretend we announce tasks on TG because it has a certain label, `announceSpecialTask` would run and the worker would not forward that because the bot would simply make the announcement as it doesn't require a user account to post.

> for example you can just define both workflow and worker endpoints in the config

Do you mean that I would define this plugin twice within `ubiquibot-config.yml` one with the worker url and one with the workflow url, on any given event both would run but would remove the self-forwarding aspect?

Reply:
> Do you mean that I would define this plugin twice within `ubiquibot-config.yml` one with the worker url and one with the workflow url, on any given event both would run but would remove the self-forwarding aspect?

Yes, exactly. I think it would simplify things and also you don't need the kernel's private key which makes it more secure.

Reply:
Sure that would remove the `APP_ID` and `PRIVATE_KEY` reliance. Considering [this comment](https://github.com/ubiquibot/telegram-bot/pull/2#discussion_r1756093851) does it make more sense to implement a kernel update to allow one plugin to define two urls i.e `-plugin: name, altUrl: url` that way future plugins don't have to be duplicated in our own and in partner configs? If the kernel reads `altUrl` it sends two payloads for that one event?

Reply:
Seems like a solution not sure if the most elegant @gentlementlegen  please decide

Reply:
Instead of having 2 entry points in the workflow, couldn't we make this as a worker which is capable to forward events to its own Actions? Not sure if I understand the problem properly.
```yml
plugins:
  - uses: https://telegram-bot.worker
    with:
      - actionCallback: "ubiquibot/telegram-bot@development"
```

Having two entries work, just makes the configuration less user friendly because it would be easy to forget one. But if that's a better solution then it's fine.

Reply:
> couldn't we make this as a worker which is capable to forward events to its own Actions

this is how I wrote it but it requires `app_private_key` which is a no-go.

> just makes the configuration less user friendly because it would be easy to forget one

50/50 which is worse, having to copy paste the entire plugin and change just the url or have to specify in the config there are two urls, whatever approach you guys think is best we'll do.

Reply:
another option is:
```yaml
plugins:
  - uses: ["https://telegram-bot.worker", "ubiquibot/telegram-bot@development"]
    with:
      - ...
```
but this might break the kernel if there is a second plugin in the chain because the kernel waits for the first plugin's response and in this case it would get 2 responses

Reply:
2 entries make sense then, just don't forget to document this in the readme.

Reply:
> 2 entries make sense then, just don't forget to document this in the readme.

So we are going to copy paste the config twice in `ubiquibot-config.yml` and the kernel will treat the worker and workflow as two separate plugins, yeah?

Reply:
you can avoid copy pasting by using [YAML anchors](https://support.atlassian.com/bitbucket-cloud/docs/yaml-anchors/)